Added Support for User Family Members

AdityaRaimec22 opened the pull request:
<!--
This section can be deleted after reading.

We employ the following branching strategy to simplify the development process and to ensure that only stable code is pushed to the `master` branch:

- `develop`: For unstable code: New features and bug fixes.
- `master`: Where the stable production ready code lies. Only security related bugs.

NOTE!!!

ONLY SUBMIT PRS AGAINST OUR `DEVELOP` BRANCH. THE DEFAULT IS `MAIN`, SO YOU WILL HAVE TO MODIFY THIS BEFORE SUBMITTING YOUR PR FOR REVIEW. PRS MADE AGAINST `MAIN` WILL BE CLOSED.

-->

<!--
Thanks for submitting a pull request! Please provide enough information so that others can review your pull request.
-->

**What kind of change does this PR introduce?**
Add Support for User Family Members
<!-- E.g. a bugfix, feature, refactoring, etc… -->

**Issue Number:**

Fixes #1584 <!--Add related issue number here.-->

**Did you add tests for your changes?**
Yes
<!--Yes or No. Note: Add unit tests or automation tests for your code.-->

**Snapshots/Videos:**

<!--Add snapshots or videos wherever possible.-->

**If relevant, did you update the documentation?**
Not relevant
<!--Add link to Talawa-Docs.-->

**Summary**

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve? -->
<!-- Try to link to an open issue for more information. -->

**Does this PR introduce a breaking change?**
No
<!-- If this PR introduces a breaking change, please describe the impact and a migration path for existing applications. -->

**Other information**
Some tests are failing and I am currently addressing them, and I have raised a pull request because the assigned issue has been pending for a significant period.
<!--Add extra information about this PR here-->

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-api/blob/master/CONTRIBUTING.md)?**
Yes
<!--Yes or No-->

palisadoes commented:
Please fix the conflicting files

palisadoes commented on tests/resolvers/Mutation/createFamilyGroup.spec.ts:
```diff
@@ -0,0 +1,76 @@
+import "dotenv/config";
+import type mongoose from "mongoose";
+import { Types } from "mongoose";
+import type { MutationCreateFamilyGroupArgs } from "../../../src/types/generatedGraphQLTypes";
```
You have interchangeably used the term `familyGroup` and `Family` in the PR definitions. To make it unambiguous for future developers use `userFamily`  to replace `familyGroup` and `Family`.

AdityaRaimec22 replied:
done

palisadoes reviewed (changes requested):
You have interchangeably used the term `familyGroup` and `Family` in the PR definitions. To make it unambiguous for future developers use `userFamily`  to replace `familyGroup` and `Family`.

palisadoes reviewed (changes requested):
The Mutation Test file names should more closely reflect the UserFamily naming convention

![image](https://github.com/PalisadoesFoundation/talawa-api/assets/16875803/823b3fdd-416f-4a3c-9946-fe0d569f1a1a)

AdityaRaimec22 commented:
> The Mutation Test file names should more closely reflect the UserFamily naming convention
> 
> ![image](https://private-user-images.githubusercontent.com/16875803/298238692-823b3fdd-416f-4a3c-9946-fe0d569f1a1a.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDU5NDA0ODEsIm5iZiI6MTcwNTk0MDE4MSwicGF0aCI6Ii8xNjg3NTgwMy8yOTgyMzg2OTItODIzYjNmZGQtNDE2Zi00YTNjLTk5NDYtZmUwZDU2OWYxYTFhLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDAxMjIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwMTIyVDE2MTYyMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ0ZGNiOTdkOGY4MGNmODg0ZDIyYTM3NTdiMmE3NzEyOTUxMmMwZTFlYTgwNTI1MmYzMWJjMTk4MWNkYjM4ZjEmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.sMncSJKMlXIMluXzImznfH4ap4VsmdPYJ6zBDa5GXuI)

committed changes @palisadoes

palisadoes commented:
Please fix the failing tests

AdityaRaimec22 commented:
working on the failing tests will fix them ASAP.

palisadoes commented:
Please fix the conflicting files

palisadoes commented:
Please fix the conflicting files

palisadoes commented:
Please fix the conflicting file

AdityaRaimec22 commented:
@palisadoes The PR is up for review now all the tests has been passed from my side.

EshaanAgg commented on src/models/User.ts:
```diff
@@ -126,6 +130,12 @@ const userSchema = new Schema(
         ref: "Organization",
       },
     ],
+    adminForUserFamily: [
+      {
+        type: Schema.Types.ObjectId,
+        ref: "Organization",
+      },
+    ],
```
The field name and the `ref` don't match. Shouldn't it be `UserFamily`?
Also, there is no need to store the same data twice, both on the User Schema and the User Family schema. We should also store it once, from where the queries and mutations would access it more frequently.

EshaanAgg commented on src/models/User.ts:
```diff
@@ -140,6 +150,12 @@ const userSchema = new Schema(
         ref: "Organization",
       },
     ],
+    createdUserFamily: [
+      {
+        type: Schema.Types.ObjectId,
+        ref: "Organization",
+      },
+    ],
```
Same comment as earlier. Remove repetition.

EshaanAgg commented on src/models/User.ts:
```diff
@@ -200,6 +216,12 @@ const userSchema = new Schema(
         ref: "Organization",
       },
     ],
+    joinedUserFamily: [
+      {
+        type: Schema.Types.ObjectId,
+        ref: "Organization",
+      },
+    ],
```
Same as above. Remove repetition.

EshaanAgg commented on src/models/userFamily.ts:
```diff
@@ -0,0 +1,56 @@
+    type: String,
+    required: true,
+  },
+  users: [
+    {
+      type: Schema.Types.ObjectId,
+      ref: "User",
+      required: true,
+    },
+  ],
+  admins: [
+    {
+      type: Schema.Types.ObjectId,
+      ref: "User",
+      required: true,
+    },
+  ],
+  creator: {
+    type: Schema.Types.ObjectId,
+    ref: "User",
+    required: true,
+  },
```
This data is repeated and adds complexity to the resolvers to maintain integrity. Either store them here or on the user object. According to me these all are better stored here, but Im open to counter arguments.

AdityaRaimec22 replied:
Committed all the asked changes the PR is up for review now. Thanks for suggestions.

EshaanAgg reviewed (changes requested):
Need some changes in the proposed schema, and subsequently, the implemented resolvers. Otherwise, the overall implementation looks good.

SiddheshKukade commented:
@AdityaRaimec22 where is the UI Equivalent fort this feature can you please link that I wanted to check that.

AdityaRaimec22 commented on src/resolvers/Mutation/adminRemoveMemberFromUserFamily.ts:
done

AdityaRaimec22 commented on src/resolvers/Mutation/adminAddMemberToUserFamily.ts:
```diff
@@ -0,0 +1,82 @@
+    });
+
+    // Checks whether user with _id === args.userId is already a member of Family.
+    if (isUserMemberOfUserFamily === true) {
```
done

AdityaRaimec22 commented on src/resolvers/Mutation/adminAddMemberToUserFamily.ts:
done

palisadoes commented:
Please fix the failing tests

AdityaRaimec22 commented:
@palisadoes @EshaanAgg All the checks are passed from my side. Can you approve the PR now. Thanks

<img width="960" alt="image" src="https://github.com/PalisadoesFoundation/talawa-api/assets/118007662/7970a2c0-f4a1-4945-9afa-e3f2aa775e1e">

AdityaRaimec22 commented:
> @AdityaRaimec22 where is the UI Equivalent fort this feature can you please link that I wanted to check that.

@SiddheshKukade here is the UI equivalent of this feature.


https://github.com/PalisadoesFoundation/talawa-api/assets/118007662/8d2ad07b-a77a-4a60-bbfd-ba91ab94ee50

palisadoes commented:
@AdityaRaimec22 

You bypassed the lint rules some how and now everyone else is having linting issues. Please resubmit your changes with the correct linting applied.

AdityaRaimec22 commented:
@palisadoesI haven't overridden or bypassed any of the changes; they were all pushed in the usual manner, and the checks also passed in this pull request. Could you please provide a link to a pull request where the lint checks failed, or share some resource indicating that checks failed for this reason? I'd like to take a look at why those changes failed.

palisadoes commented:
@AdityaRaimec22 

1. That was the impression given to me by @LakshyaSatpal by this comment right after i merged. When it was reverted the issue went away.
2. However, that may have been caused by fixing merge conflicts which may have triggered the problem, based on the fact that he mentioned the conflicts going away, which were not mentioned initially.  It could have been a question of timing as it was noted very soon after the merge.
    1. https://github.com/PalisadoesFoundation/talawa-api/pull/1774#issuecomment-1925459699
If you feel confident that the linting was OK, let's resubmit and let's try to see whether others have the same issue.

I'll hang around until you do.

palisadoes commented:
If I was hasty, I apologize. The timing was so close that I might have jumped to an incorrect conclusion

AdityaRaimec22 commented:
> If I was hasty, I apologize. The timing was so close that I might have jumped to an incorrect conclusion

@palisadoes so should I open a new PR.

palisadoes commented:
Yes

LakshyaSatpal commented:
@AdityaRaimec22 In most files, the use of `any` in your changes, was causing lint errors.

AdityaRaimec22 commented:
> @AdityaRaimec22 In most cases, the use of `any` in your changes, was causing lint errors. I confirmed this by looking at the changed files in this PR.

@LakshyaSatpal The code I've added isn't failing in the checks of this PR or when tested locally. Additionally, the code I've implemented draws inspiration from existing code in the codebase; I haven't introduced any new methods to incorporate this feature. It is due to other reasons which we are working on.
